src/api/spotify-api.ts

```diff
@@ -0,0 +1,54 @@
+import { refreshAccessToken } from '../main/auth';
+
+const API_URL = 'https://api.spotify.com/v1';
+
+class SpotifyApi {
+  private isThrottled: boolean;
+  private throttleTime: number;
+
+  public refreshToken: string;
+
+  async fetch(input: RequestInfo, init?: RequestInit) {
+    if (this.isThrottled) {
+      const timeLeft = this.throttleTime - new Date().getTime();
+      if (timeLeft > 0) {
+        console.warn(
+          `Spotify API call ignored, waiting ${timeLeft}ms... (${input})`
+        );
+        return;
+      } else {
+        this.isThrottled = false;
+      }
+    }
+
+    const res = await fetch(API_URL + input, init);
+    switch (res.status) {
+      case 200: {
+        return await res.json();
+      }
+      case 204: {
+        return null;
+      }
+      case 401: {
+        if (this.refreshToken) {
+          await refreshAccessToken(this.refreshToken);
+        }
+      }
+      case 429: {
+        const retryAfter = parseInt(res.headers.get('retry-after')) + 1;
+        if (retryAfter) {
+          this.throttleTime = new Date().getTime() + retryAfter * 1000;
+          this.isThrottled = true;
+        }
+        break;
+      }
+    }
+
+    const { error } = await res.json();
+    console.error(`${error.status}: ${error.message}`);
+
+    return null;
```

Comment:
I think this should be throwing an error instead of returning null (it's not the api's responsibility to handle the error, even if it's only via a console.error). Curious as to what your thoughts are on that (it might not be worth the extra effort since this 'api' is super specific to Spotify/Lofi

Reply:
Yeah, basically it depends where the error should be handled. If we want to handle it downstream, we can throw; if not, returning `null` is fine. In this case, I agree with you that the scope is so limited to Spotify/Lofi that returning null should be fine. I'd probably just document it so if anyone ever looks at the code, they understand our rationale.